File: sdks/canary/typescript/scenarios/s-apikey.ts
// Copyright (C) 2026 Michael Kao
// SPDX-License-Identifier: AGPL-3.0-or-later
// S-APIKEY canary — TypeScript SDK

import { LLMSafeSpaces } from '../../../typescript/src/index.js';
import { Runner, Config, configFromEnv, nodeFetch } from '../canary.js';

async function run(r: Runner, cfg: Config): Promise<void> {
  const c = new LLMSafeSpaces({ baseUrl: cfg.apiUrl, apiKey: cfg.apiKey, timeout: 20000, fetch: nodeFetch as any });
  let keyId: string | null = null;
  try {
    const [ok, key] = await r.assertNoError(() => c.auth.createApiKey('canary-ts-key'), 'create-key: no error');
    if (!ok || !key) return;
    r.assert(key.name === 'canary-ts-key', 'create-key: name');
    r.assert(key.key?.startsWith('lsp_') ?? false, 'create-key: starts with lsp_', key.key);
    r.assert(key.active === true, 'create-key: active=true');
    keyId = key.id;

    // List — key present, full key absent
    const [ok2, keys] = await r.assertNoError(() => c.auth.listApiKeys(), 'list-keys: no error');
    if (ok2 && keys) {
      const found = (keys as any[]).find((k: any) => k.id === keyId);
      r.assert(!!found, 'list-keys: key present');
      if (found) r.assert(!found.key, 'list-keys: full key absent');
    }

    // New key authenticates
    const nc = new LLMSafeSpaces({ baseUrl: cfg.apiUrl, apiKey: key.key, timeout: 10000, fetch: nodeFetch as any });
    await r.assertNoError(() => nc.auth.me(), 'new-key: authenticates');

    // Delete
    await r.assertNoError(() => c.auth.deleteApiKey(keyId!), 'delete-key: no error');
    keyId = null;

    // Absent after delete
    const [ok3, keys2] = await r.assertNoError(() => c.auth.listApiKeys(), 'list-after-delete: no error');
    if (ok3 && keys2) r.assert(!(keys2 as any[]).some((k: any) => k.id === key.id), 'list-after-delete: absent');

    // Deleted key rejected
    await r.assertError(() => nc.auth.me(), 'deleted-key: AuthError');

  } finally {
    if (keyId) { try { await c.auth.deleteApiKey(keyId); } catch {} }
  }
  await r.assertError(() => c.auth.deleteApiKey('00000000-0000-0000-0000-000000000099'), 'delete-nonexistent: error');
}

async function main() {
  const r = new Runner('apikey');
  await run(r, configFromEnv());
  r.print();
  process.exit(r.exitCode());
}
main().catch(e => { console.error(e); process.exit(1); });

File: sdks/canary/python/scenarios/s_apikey.py
#!/usr/bin/env python3
# Copyright (C) 2026 Michael Kao
# SPDX-License-Identifier: AGPL-3.0-or-later
"""S-APIKEY canary — Python SDK"""

from __future__ import annotations
import sys, os

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), ".."))
from canary import Runner, Config, config_from_env
from llmsafespaces import LLMSafeSpaces


def run(r: Runner, cfg: Config) -> None:
    c = LLMSafeSpaces(cfg.api_url, api_key=cfg.api_key, timeout=20.0)
    key_id = None
    try:
        ok, key = r.assert_no_error(
            lambda: c.auth.create_api_key("canary-py-key"), "create-key: no error"
        )
        if not ok:
            return
        r.assert_(key.name == "canary-py-key", "create-key: name")
        r.assert_(key.key.startswith("lsp_"), "create-key: starts with lsp_", key.key)
        r.assert_(key.active, "create-key: active=true")
        key_id = key.id

        # List — key present, full key absent
        ok2, keys = r.assert_no_error(
            lambda: c.auth.list_api_keys(), "list-keys: no error"
        )
        if ok2:
            found = next((k for k in keys if k.id == key_id), None)
            r.assert_(found is not None, "list-keys: key present")
            if found:
                r.assert_(not found.key, "list-keys: full key absent in list")

        # New key authenticates
        new_c = LLMSafeSpaces(cfg.api_url, api_key=key.key, timeout=10.0)
        r.assert_no_error(lambda: new_c.auth.me(), "new-key: authenticates")

        # Delete
        r.assert_no_error(lambda: c.auth.delete_api_key(key_id), "delete-key: no error")
        key_id = None

        # Absent after delete
        ok3, keys2 = r.assert_no_error(
            lambda: c.auth.list_api_keys(), "list-after-delete: no error"
        )
        if ok3:
            r.assert_(all(k.id != key.id for k in keys2), "list-after-delete: absent")

        # Deleted key rejected
        r.assert_error(lambda: new_c.auth.me(), "deleted-key: AuthError")

    finally:
        if key_id:
            try:
                c.auth.delete_api_key(key_id)
            except:
                pass

    # N1: delete nonexistent
    r.assert_error(
        lambda: c.auth.delete_api_key("00000000-0000-0000-0000-000000000099"),
        "delete-nonexistent: error",
    )


if __name__ == "__main__":
    r = Runner("apikey")
    cfg = config_from_env()
    run(r, cfg)
    r.print()
    sys.exit(r.exit_code())
